File: kafka-editor/src/codegen/topoSort.ts
/**
 * Topological sort of the topology graph (Kahn's algorithm). Sources come
 * first, so codegen can emit each node after the nodes it depends on.
 */
export function topoSort<N extends { id: string }>(
  nodes: N[],
  edges: { source: string; target: string }[],
): N[] | null {
  const indegree = new Map<string, number>()
  const adjacency = new Map<string, string[]>()
  for (const node of nodes) {
    indegree.set(node.id, 0)
    adjacency.set(node.id, [])
  }
  for (const edge of edges) {
    if (!indegree.has(edge.source) || !indegree.has(edge.target)) continue
    adjacency.get(edge.source)?.push(edge.target)
    indegree.set(edge.target, (indegree.get(edge.target) ?? 0) + 1)
  }

  const queue = nodes
    .filter((n) => (indegree.get(n.id) ?? 0) === 0)
    .map((n) => n.id)
  const order: string[] = []
  while (queue.length > 0) {
    const id = queue.shift() as string
    order.push(id)
    for (const next of adjacency.get(id) ?? []) {
      const remaining = (indegree.get(next) ?? 0) - 1
      indegree.set(next, remaining)
      if (remaining === 0) queue.push(next)
    }
  }

  // A short order means a cycle left some nodes unreachable.
  if (order.length !== nodes.length) return null

  const byId = new Map(nodes.map((n) => [n.id, n]))
  return order.flatMap((id) => {
    const node = byId.get(id)
    return node ? [node] : []
  })
}

File: kafka-editor/backend/src/main/java/se/afshin/yavari/kafka/editor/interpreter/TopoSort.java
package se.afshin.yavari.kafka.editor.interpreter;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import se.afshin.yavari.kafka.editor.model.ProjectEdge;
import se.afshin.yavari.kafka.editor.model.ProjectNode;

/** Topological sort of the topology graph (Kahn's algorithm). */
public final class TopoSort {

    private TopoSort() {
    }

    /** Returns the nodes in dependency order. Throws if the graph has a cycle. */
    public static List<ProjectNode> sort(List<ProjectNode> nodes, List<ProjectEdge> edges) {
        Map<String, Integer> indegree = new HashMap<>();
        Map<String, List<String>> adjacency = new HashMap<>();
        Map<String, ProjectNode> byId = new HashMap<>();
        for (ProjectNode node : nodes) {
            indegree.put(node.id(), 0);
            adjacency.put(node.id(), new ArrayList<>());
            byId.put(node.id(), node);
        }
        for (ProjectEdge edge : edges) {
            if (!indegree.containsKey(edge.source()) || !indegree.containsKey(edge.target())) {
                continue;
            }
            adjacency.get(edge.source()).add(edge.target());
            indegree.merge(edge.target(), 1, Integer::sum);
        }

        Deque<String> queue = new ArrayDeque<>();
        for (ProjectNode node : nodes) {
            if (indegree.get(node.id()) == 0) {
                queue.add(node.id());
            }
        }
        List<ProjectNode> order = new ArrayList<>();
        while (!queue.isEmpty()) {
            String id = queue.poll();
            order.add(byId.get(id));
            for (String next : adjacency.get(id)) {
                if (indegree.merge(next, -1, Integer::sum) == 0) {
                    queue.add(next);
                }
            }
        }
        if (order.size() != nodes.size()) {
            throw new IllegalArgumentException("The topology contains a cycle");
        }
        return order;
    }
}
